src/Action/Generate.hs

```diff
@@ -126,17 +128,25 @@ readHaskellOnline timing settings download = do
     return (cbl, want, source)
 
 
-readHaskellDirs :: Timing -> [FilePath] -> IO (Map.Map String Package, Set.Set String, Source IO (String, URL, LStr))
-readHaskellDirs timing dirs = do
+readHaskellDirs :: Timing -> Settings -> [FilePath] -> IO (Map.Map String Package, Set.Set String, Source IO (String, URL, LStr))
+readHaskellDirs timing settings dirs = do
     packages <- map (takeBaseName &&& id) . filter ((==) ".txt" . takeExtension) <$> concatMapM listFilesRecursive dirs
+    cabals <- concatMapM listFilesRecursive dirs >>=
```

Comment:
The listFilesRecursive should be done once and shared breed packages and cabals, since it can be relatively expensive.